Application/EdFi.Ods.Features/Publishing/ExceptionHandling/SnapshotGoneExceptionTranslator.cs

```diff
@@ -0,0 +1,59 @@
+using System;
+using System.Net;
+using EdFi.Common;
+using EdFi.Common.Extensions;
+using EdFi.Ods.Api.ExceptionHandling;
+using EdFi.Ods.Api.Exceptions;
+using EdFi.Ods.Api.Models;
+using EdFi.Ods.Extensions.Publishing.Feature.SnapshotContext;
+using NHibernate.Exceptions;
+
+namespace EdFi.Ods.Extensions.Publishing.Feature.ExceptionHandling
+{
+    /// <summary>
+    /// Implements an exception translator that looks for the custom <see cref="DatabaseConnectionException" />
+    /// embedded in an <see cref="GenericADOException" />, and if snapshot context has been provided on
+    /// the request, sets the response status code to HTTP 410 (Gone).
```

Comment:
Well, the original intention was that each snapshot identifier would be GUID-based and would not re-appear later as a different snapshot (the GUID-length restriction has been subsequently dropped to avoid issues with database name length, particularly with Postgres and other community contributions).

So, I think the `410` status is consistent with the original _intent_, there is nothing we can do (other than documenting the assumption in some feature documentation) to stop implementers from reusing snapshot identifiers.

If we changed this to a `404` response, would it be distinguishable from a `404` response that is happening because the resource actually doesn't exist? For example, when the API client is trying to access an extension resource that actually doesn't exist on the source API? (The API Publisher uses the target API's dependency metadata to drive its processing -- but a source API might not have the same extension installed and `404`s would result in that scenario.)

Perhaps it would come down to the API client establishing what resources are available from the source, and hoping the snapshot doesn't get pulled in the middle of that initialization process? Either that, or it'd have to check the message body on the `404` response.

Personally, I think the `410` is appropriate here. With that context, what are your thoughts?

Reply:
I'm less concerned on reusing the snapshot identifier, and more concerned with a client trying to get the snapshot before its ready, and the implications of Gone.  I see the concerns on using 404 since this is baked into the header, would something like 424 Failed Dependency be a good fit here?

Reply:
The snapshot record in the `publishing.snapshot` table shouldn't be written by the host until _after_ the database snapshot has been created and is available. That implementation is handled out-of-band from this, but I think that should be the expectation for implementers of this -- create your database snapshot (or restore your backup) and then write the record in the table when that completes. When removing a snapshot or backup database, drop the corresponding `publishing.snapshot` record in the ODS first, then remove the database (so that new requests from API clients don't result in them trying to start using it).

On Mozilla's site, they document `424 Failed Dependency` as part of WebDAV with a description of:

> The request failed due to failure of a previous request.

That's not really what's happening - there was no previous failing request that led to this. This literally is a case where the snapshot of the resource you were looking for is gone (the underlying database no longer exists) and it is never coming back.

Reply:
To play devil's advocate with an edge scenario... the 410 is based on a failure to connect to the database based on the snapshot header. If there was some sort of transient issue affecting connectivity to the database server, the snapshot response might falsely state it's "Gone". Then, not sure how the aforementioned caching might affect the API client if it re-retrieved the latest snapshot from the _/publishing/snapshots_ endpoint and then ultimately tried the resource again. If we stick with `410`, would we want to add a `Cache-Control: no-store` header to the response to prevent this from being problematic?

Reply:
I think a lot of this is a side effect of using headers to drive addressing the historical versions of these resources, rather than using query string or route.  If this was a unique url then the 410 would be less concerning, because yes, it really is gone forever at that url, but since its just a difference of headers caching can be a little weirder.  Since this is work coming off of another work stream I'm assuming that decision is fairly locked in at this point.  So given the usual usage of this api not being on a browser (which is where a lot of these caching concerns would surface anyways), I think if we just add the Cache-Control as you mentioned that's enough for here, implementations coding against this would respect expectations of this usage and not assume 410 means gone forever.

Reply:
I'm curious about how you would go about implementing a feature that needs to change the routes on existing controllers (i.e. to introduce snapshot into the route of the API), particularly now that the route metadata is attribute-based on the controllers? This feature just needs to inject a snapshot identifier into the database name during construction of the connection string. I don't think it's an irreversible design decision at this point. I would just need to update the API Publisher to know about this change, and as far as I know right now, it's only Oregon Nexus who is using it regularly.

Reply:
Using something like we do for DataManagementControllerRouteConvention and managing whether the school year is part of the route or not.  The resource route can be stable, and the route convention can transform all relevant routes in a consistent manner, and still have it point to the right controller.  Then a middleware layer to set the snapshot context, same as the school year is handled.  If the route was unique I would have zero concerns with sending a 410, because in that case it really is gone based on the way the feature works and relative to that specific route.  Like I said though, the reality is the concerns are probably minimal for caching going on particularly if you set the cache header, so its likely just down to a style preference, I usually prefer route based over header/configuration based for driving what version of a resource you're accessing.

Reply:
How would that approach impact the Open API metadata we produce?

Reply:
@blmeyers If the route is modified to include snapshot, wouldn't it complicate routes (specially with instance and year specific routes added recently)

Reply:
Yeah, the route would end up being a lot and since it doesn't really mix with the other routes, the optimal would probably be to make it query string driven, similar to how change queries work.  A url with a different query string is still considered "unique".  Still doesn't solve the issue there with differentiating 404s for missing extension versus missing snapshot, but I'm not sure there is a great way to handle that other than a different status code and assuming the difference.

Regarding OpenApiMetadata, that's actually an area we need a better seam for features that modify the way all other data standard resources act, such as Change Queries.  Right now Change Queries has it "baked in" to the default code and has a feature check in there, which isn't a good example for other features.

I think based on the way the field tends to use these features that we probably would be ok without modifying this, and keeping the 410 Gone.  That will allow more savvy clients to have at least some mechanism to differentiate between the snapshot missing and resource missing scenarios.

So if we stick with 410 Gone, are we going to update the Open Api Metadata to add this code and communicate what that status code means in the context of the resource endpoints?